crates/goose/src/agent.rs

```diff
@@ -127,77 +133,176 @@ impl Agent {
+                            *token_count
+                        ));
+                    }
+                }
+            }
+
+            // Sort by priority (high to low) and timestamp (newest to oldest)
+            all_resources.sort_by(|a, b| {
+                let priority_cmp = b.2.priority.cmp(&a.2.priority);
+                if priority_cmp == std::cmp::Ordering::Equal {
+                    b.2.timestamp.cmp(&a.2.timestamp)
+                } else {
+                    priority_cmp
+                }
+            });
+
+            // Remove resources until we're under target limit
+            let mut current_tokens = approx_count;
+
+            while current_tokens > target_limit && !all_resources.is_empty() {
+                if let Some((system_name, uri, _, token_count)) = all_resources.pop() {
+                    if let Some(system_counts) = system_token_counts.get_mut(&system_name) {
+                        system_counts.remove(&uri);
+                        current_tokens -= token_count as usize;
+                    }
+                }
+            }
+            // Create status messages only from resources that remain after token trimming
+            for (system_name, uri, _, _) in &all_resources {
+                if let Some(system_resources) = resource_content.get(system_name) {
+                    if let Some((resource, content)) = system_resources.get(uri) {
+                        status_content.push(format!("{}\n```\n{}\n```\n", resource.name, content));
+                    }
+            }
+        }
+        }
+        else {
+            // Create status messages from all resources when no trimming needed
+            for (_system_name, resources) in &resource_content {
+                for (resource, content) in resources.values() {
+                    status_content.push(format!("{}\n```\n{}\n```\n", resource.name, content));
+                }
+            }
+        }
+
+        // Join remaining status content and create status message
+        let status_str = status_content.join("\n");
+        let mut context = HashMap::new();
+        let systems_status = vec![SystemStatus::new("system", status_str)];
+        context.insert("systems", &systems_status);
+
+        // Load and format the status template with only remaining resources
+        let status = load_prompt_file("status.md", &context)
+            .map_err(|e| AgentError::Internal(e.to_string()))?;
+
+
+        // Create a new messages vector with our changes
+        let mut new_messages = messages.to_vec();
+
+        // Add pending messages
+        for msg in pending {
+            new_messages.push(msg.clone());
+        }
+
+        // Finally add the status messages
         let message_use =
             Message::assistant().with_tool_request("000", Ok(ToolCall::new("status", json!({}))));
 
         let message_result =
             Message::user().with_tool_response("000", Ok(vec![Content::text(status)]));
```

Comment:
if we append the status of the tools in the exchange, is this going to actually use up quite a bit of the context window? or are the statuses only of the called tools (not all tools/systems)